tee-worker/litentry/core/assertion-build/src/a14.rs

```diff
@@ -0,0 +1,210 @@
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+use crate::*;
+use frame_support::traits::Len;
+use itp_stf_primitives::types::ShardIdentifier;
+use itp_types::AccountId;
+use itp_utils::stringify::account_id_to_string;
+use lc_credentials::Credential;
+use lc_data_providers::{
+	graphql::{AchainableQuery, GraphQLClient, VerifiedCredentialsIsHodlerIn},
+	vec_to_string,
+};
+use litentry_primitives::{Address32, SupportedNetwork};
+use log::*;
+use std::{
+	format,
+	string::{String, ToString},
+	vec::Vec,
+};
+
+const VC_SUBJECT_DESCRIPTION: &str =
+	"The user has been consistently holding at least {x} amount of {y} tokens at given years on native network";
+const VC_SUBJECT_TYPE: &str = "Native Token Holding Assertion";
+const VC_SUBJECT_TAG: [&str; 2] = ["Litentry", "Polkadot"];
+
+pub fn build(
+	min_balance: ParameterString,
+	years: ParameterYears,
+	token: NativeToken,
+	identities: Vec<Identity>,
+	shard: &ShardIdentifier,
+	who: &AccountId,
+) -> Result<Credential> {
+	debug!(
+		"Assertion A14 build, who: {:?}, identities: {:?}",
+		account_id_to_string(&who),
+		identities
+	);
+
+	let q_min_balance = vec_to_string(min_balance.to_vec()).map_err(|_| {
+		Error::RequestVCFailed(
+			Assertion::A14(min_balance.clone(), token.clone(), years.clone()),
+			ErrorDetail::ParseError,
+		)
+	})?;
+
+	let token_str = vec_to_string(token.as_string().to_vec()).map_err(|_| {
+		Error::RequestVCFailed(
+			Assertion::A14(min_balance.clone(), token.clone(), years.clone()),
+			ErrorDetail::ParseError,
+		)
+	})?;
+
+	let mut client = GraphQLClient::new();
+	let supported_network: SupportedNetwork = token.clone().into();
+	let addresses: Vec<String> = collect_addresses_from_identities(identities, supported_network);
+
+	if addresses.is_empty() {
+		return Err(Error::RequestVCFailed(
+			Assertion::A14(min_balance, token, years),
+			ErrorDetail::StfError(ErrorString::truncate_from(
+				format!("Account {:?} has no addresses on requested networks", who)
+					.as_bytes()
+					.to_vec(),
+			)),
+		))
+	}
+
+	let assertion_dates = prepare_assertion_dates(&years).map_err(|_| {
+		Error::RequestVCFailed(
+			Assertion::A14(min_balance.clone(), token.clone(), years.clone()),
+			ErrorDetail::ParseError,
+		)
+	})?;
+	let mut results: Vec<(String, bool)> = Vec::with_capacity(assertion_dates.len());
+	for from_date in assertion_dates.iter() {
+		let vch = VerifiedCredentialsIsHodlerIn::new(
```

Comment:
I assume this request will answer the question wheter the given address was holding a token since `from_date` to end of that year. Can anybody confirm ?